### ISSUE TITLE
ResourceManager doesn't properly support non string resources on .NET Core msbuild

### ISSUE THREAD

**Issue by alexperovich** [dotnet/corefx#12565]
The following tests fail on Unix platforms:
- System.Resources.Tests.ResourceManagerTests.GetObject(key: \"Int\", expectedValue: 42)
- System.Resources.Tests.ResourceManagerTests.GetObject(key: \"Float\", expectedValue: 3.14159)
- System.Resources.Tests.ResourceManagerTests.GetObject(key: \"Bytes\", expectedValue: [41, 42, 43, 44, 192, ...])
- System.Resources.Tests.ResourceManagerTests.GetStream
- System.Resources.Tests.ResourceManagerTests.GetResourceSet_NonStrings(key: \"Int\", expectedValue: 42)
- System.Resources.Tests.ResourceManagerTests.GetResourceSet_NonStrings(key: \"Float\", expectedValue: 3.14159)
- System.Resources.Tests.ResourceManagerTests.GetResourceSet_NonStrings(key: \"Bytes\", expectedValue: [41, 42, 43, 44, 192, ...])

See #12468

**Comment by AlexGhiondea** [comment]
@alexperovich can you work with @tarekgh on this?

**Comment by alexperovich** [comment]
This is caused by an underlying msbuild issue where the GenerateResources task doesn't honor the types in the resx when generating resources. Microsoft/msbuild#1352

**Comment by tarekgh** [comment]
@alexperovich if this is msbuild issue then why this is happening on Linux only and not happening on Windows?

**Comment by alexperovich** [comment]
Its actually a .NET Core msbuild issue. I will change the title. It doesn't happen on windows because we use full .NET msbuild there.

**Comment by danmosemsft** [comment]
I'm going to close this since there's no action to take on our part. Does seem too important whether we take the active issue off the tests, they're correctly disabled on Unix.

**Comment by Jiayili1** [comment]
Disabled test, reopen it.

**Comment by Jiayili1** [comment]
https://github.com/dotnet/corefx/blob/master/src/System.Resources.ResourceManager/tests/ResourceManagerTests.netstandard.cs#L86

**Comment by ViktorHofer** [comment]
@Jiayili1 the link you provided does not work.

**Comment by alexperovich** [comment]
The new link would be here: https://github.com/dotnet/corefx/blob/master/src/System.Resources.ResourceManager/tests/ResourceManagerTests.cs#L86

**Comment by ViktorHofer** [comment]
Thanks Alex. Is the msbuild issue you mention also relevant for the sources or only for the tests above? So does it need to be fixed till 2.0 milestone?

**Comment by alexperovich** [comment]
Any resx that has resources in it that aren't strings will work incorrectly when build with .net core msbuild. It only shows up in these tests because these tests are intended to test that scenario.

**Comment by gkhanna79** [comment]
@tarekgh @danmosemsft  Is this expected to work .NET Core 2.0?

**Comment by danmosemsft** [comment]
I don't think so, there's been no movement on https://github.com/Microsoft/msbuild/issues/1352 and we've not asked for it.

**Comment by tarekgh** [comment]
just wondering, looks this working for Windows build. is it right? would be strange to have this working for Windows and not for Linux.

**Comment by ViktorHofer** [comment]
Do we have resources which aren't pure strings? If yes then this is a *must* for 2.0

**Comment by tarekgh** [comment]
The framework doesn't have non string resources (so far) but the app can have that.

**Comment by gkhanna79** [comment]
I am not aware of .NET Core apps having non-string resources embedded in them - @tarekgh how would they get it?

**Comment by tarekgh** [comment]
There are 2 scenarios here: the first is you can embed any file inside your assembly resources using EmbeddedResource msbuild element and you can read this embedded file at runtime using Assembly.GetManifestResourceStream. this scenario should be already supported today on Windows and Linux because we already using this in some of our assemblies. 

The second scenario is the app creates resx files has embedded files/data like the example https://msdn.microsoft.com/en-us/library/963f81yd(v=vs.110).aspx then the app can read the data using ResourceManager.GetObject API. I am not sure if the net core app tooling can allow people create such resx file.  Looks from the the test we already have it is possible to run into similar scenario. My whole point here is to make sure Windows behavior exactly same as Linux behavior.

**Comment by danmosemsft** [comment]
If this is important to customers we should make that case to MSBuild -- so far as I know there is no work on CoreFX right?

**Comment by ViktorHofer** [comment]
Exactly, the issue which already exists in coreclr (https://github.com/Microsoft/msbuild/issues/1352) should be prioritized. Nothing to do here...

**Comment by tarekgh** [comment]
If msbuild didn't fix this for v2.0, should we make the APIs using non string resource to throw on all platforms? so we can have consistent story?

**Comment by danmosemsft** [comment]
My take -- that's probably not the pattern we wish to follow (we would prefer to bring all functionality together, but for many folks Unix is not on the radar and this is valuable to them)

Do we have data on the need (eg., specific asks) that we could use to support our ask on MSBuild?

**Comment by tarekgh** [comment]
we have been not supporting non string resources for awhile in UWP and I didn't see any real complain but such apps had the choice to use WinRT model to achieve such scenario.

We still have a way for the app to embed files using EmbeddedResource element. The drawback of this approach is the resources will be neutral resources (cannot be associated with any localized language) and also the apps cannot do automatic type casting from the resources to specific type.

In short, I didn't see any strong request to support this scenario so far. this scenario will be much valuable when we start supporting UI stack though.

**Comment by ViktorHofer** [comment]
I think supporting non string resources is a basic requirement for any king of Resource Manager implementation. Sure - as Tarek said - it will be much more important when UI stacks are joining the party. I think we shouldn't wait for that and I also see a few (server-side) uses-cases in aspnetcore for that.

**Comment by danmosemsft** [comment]
we are past feature complete date -- it would need dev mgr approval at this point and they would need good rationale (backing data why we should do this instead of work on bugs and perf...). I think it is unlikely.

**Comment by danmosemsft** [comment]
I think the work here is basically to bring back BinaryFormatter in ResourceReader. Since since this bug was opened, we ported it to .NET Core.
https://referencesource.microsoft.com/#mscorlib/system/resources/resourcereader.cs,107
I am not sure why int, etc would fail as stated in the top post. That path does exist.

@tarekgh @ViktorHofer does that sound like the only work? Maybe @alexperovich  you can remember what the issue was with int.

**Comment by ViktorHofer** [comment]
> I think the work here is basically to bring back BinaryFormatter in ResourceReader

If my memory serves me right, we aren't using binaryformatter anywhere inside the BCL. That's intentional because of the pitfalls we already discussed various times. IMO adding BinaryFormatter back in ResourceReader isn't ideal but I don't have comprehensive knowledge about ResourceManager. Are we investing in RM / is it a straight port, or are we working on a new version of RM?
cc @stephentoub

**Comment by danmosemsft** [comment]
We do use BinaryFormatter in several places in the BCL. As long as it is used for compat like this, and not finding new uses for it. 

The question of course is whether this feature is interesting given our fixed set of serializable types (plus any custom types). Perhaps we have insufficient motivation.

Some relevant issues - 
https://github.com/Microsoft/msbuild/issues/1352 
https://github.com/Microsoft/msbuild/issues/2248

**Comment by tarekgh** [comment]
@danmosemsft I am not sure why we think this is a BinaryFormatter issue? per @alexperovich is msbuild issue https://github.com/Microsoft/msbuild/issues/1352. how did you link this to the BinaryFormatter?

**Comment by danmosemsft** [comment]
@tarekgh I misunderstood this issue. I see it is merely tracking waiting for the MSBuild fix to Microsoft/msbuild#1352 to work around the currently missing ResXResourceReader (or consume it if it is ever ported).

It seems we need a new issue then for the fact that our ResourceReader and ResourceWriter do not support binary formatted blobs in resources. It may be that it is not very useful for them to do so until ResXResourceReader exists also.

**Comment by tarekgh** [comment]
@nguerrera is this related to https://github.com/Microsoft/msbuild/issues/2221?

**Comment by tarekgh** [comment]
Microsoft/msbuild#2221
